src/couch/test/eunit/couch_js_tests.erl

```diff
@@ -33,15 +22,101 @@ couch_js_test_() ->
             fun test_util:start_couch/0,
             fun test_util:stop_couch/1,
             [
+                fun should_create_sandbox/0,
+                fun should_roundtrip_utf8/0,
+                fun should_roundtrip_modified_utf8/0,
+                fun should_replace_broken_utf16/0,
                 {timeout, 60000, fun should_exit_on_oom/0}
             ]
         }
     }.
 
 
+should_create_sandbox() ->
+    % Try and detect whether we can see out of the
+    % sandbox or not.
+    Src = <<
+      "function(doc) {\n"
+      "  try {\n"
+      "    emit(false, typeof(Couch.compile_function));\n"
+      "  } catch (e) {\n"
+      "    emit(true, e.message);\n"
+      "  }\n"
+      "}\n"
+    >>,
+    Proc = couch_query_servers:get_os_process(<<"javascript">>),
+    true = couch_query_servers:proc_prompt(Proc, [<<"add_fun">>, Src]),
+    Result = couch_query_servers:proc_prompt(Proc, [<<"map_doc">>, <<"{}">>]),
+    ?assertEqual([[[true, <<"Couch is not defined">>]]], Result).
+
+
+should_roundtrip_utf8() ->
+    % Try round tripping UTF-8 both directions through
+    % couchjs. These tests use hex encoded values of
+    % Ä (C384) and Ü (C39C) so as to avoid odd editor/Erlang encoding
+    % strangeness.
+    Src = <<
+      "function(doc) {\n"
+      "  emit(doc.value, \"", 16#C3, 16#9C, "\");\n"
+      "}\n"
+    >>,
+    Proc = couch_query_servers:get_os_process(<<"javascript">>),
+    true = couch_query_servers:proc_prompt(Proc, [<<"add_fun">>, Src]),
+    Doc = {[
+        {<<"value">>, <<16#C3, 16#84>>}
+    ]},
+    Result = couch_query_servers:proc_prompt(Proc, [<<"map_doc">>, Doc]),
+    ?assertEqual([[[<<16#C3, 16#84>>, <<16#C3, 16#9C>>]]], Result).
+
+
+should_roundtrip_modified_utf8() ->
+    % Mimicing the test case from the mailing list
+    Src = <<
+      "function(doc) {\n"
+      "  emit(doc.value.toLowerCase(), \"", 16#C3, 16#9C, "\");\n"
+      "}\n"
+    >>,
+    Proc = couch_query_servers:get_os_process(<<"javascript">>),
+    true = couch_query_servers:proc_prompt(Proc, [<<"add_fun">>, Src]),
+    Doc = {[
+        {<<"value">>, <<16#C3, 16#84>>}
+    ]},
+    Result = couch_query_servers:proc_prompt(Proc, [<<"map_doc">>, Doc]),
+    ?assertEqual([[[<<16#C3, 16#A4>>, <<16#C3, 16#9C>>]]], Result).
+
+
+should_replace_broken_utf16() ->
+    % This test reverse the surrogate pair of
+    % the Boom emoji U+1F4A5
```

Comment:
:boom:

Reply:
FYI this fails for me locally in my SM68 work, when I copy over just your test. The others succeed.

Reply:
Can you paste the failure? Might be obvious if I can see the binary mismatch.

Reply:
```
module 'couch_js_tests'
  Test couchjs
    couch_js_tests: couch_js_test_...[0.052 s] ok
    couch_js_tests: couch_js_test_...[0.049 s] ok
    couch_js_tests: couch_js_test_...[0.043 s] ok
    couch_js_tests: couch_js_test_...*failed*
in function couch_util:json_decode/2 (src/couch_util.erl, line 508)
in call from couch_js_tests:should_replace_broken_utf16/0 (test/eunit/couch_js_tests.erl, line 101)
**throw:{invalid_json,{11,invalid_string}}
  output:<<"">>

    couch_js_tests: couch_js_test_...out of memory
[1.280 s] ok
```

Note I haven't yet pulled over your removal-of-UTF8 changes, so it might be a bug somewhere in the `enc_/dec_` functions I haven't bothered to diagnose.

Reply:
There's a smell off-by-one error in that the old code would expect to see a trailing `\0` null byte that no longer goes. I've had to tweak things a few times to make sure that I've got it right.